### PR TITLE
Ability to use vanilla pytorch instead of tcnn for Nerfacto 

### PR DESCRIPTION
**Pull request by ethanweber**
- [x] `--pipeline.model.implementation tcnn` is default but `--pipeline.model.implementation torch` can be used for vanilla pytorch. This will only work for the Nerfacto method at the moment.
- [x] Removed "nerfacto" from the blacklisted github actions tests.

The vanilla pytorch implementation is maybe ~4x slower. Looking into numerical equivalence is a good next step.

Example commands are as follows,

```
ns-train nerfacto --vis viewer+wandb --experiment-name aspen-tcnn --data data/nerfstudio/aspen --pipeline.model.implementation tcnn
```

and

```
ns-train nerfacto --vis viewer+wandb --experiment-name aspen-torch --data data/nerfstudio/aspen --pipeline.model.implementation torch
```

The quality looks very similar.

The main branch and this PR (which is "tcnn" here and still the default for nerfacto) should be equivalent. Here is a small sanity check.

image

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -71,6 +67,13 @@ gen = [
     "transformers==4.24.0",
 ]
 
+# Geometry export related dependencies
+export = [
+    "open3d>=0.16.0",
+    "pymeshlab>=2022.2.post2",
+    "trimesh>=3.20.2"
```
Comment (tancik) [inline pyproject.toml:74]: missing comma

File: nerfstudio/field_components/encodings.py [hunk 2]
```diff
@@ -118,6 +124,20 @@ def __init__(
         self.max_freq = max_freq_exp
         self.include_input = include_input
 
+        self.tcnn_encoding = None
+        if implementation == "torch":
+            pass
```
Comment (tancik) [inline nerfstudio/field_components/encodings.py:129]: Remove

File: nerfstudio/field_components/encodings.py [hunk 3]
```diff
@@ -647,17 +676,38 @@ class SHEncoding(Encoding):
...
         super().__init__(in_dim=3)
 
         if levels <= 0 or levels > 4:
             raise ValueError(f"Spherical harmonic encoding only supports 1 to 4 levels, requested {levels}")
 
         self.levels = levels
 
+        self.tcnn_encoding = None
+        if implementation == "torch":
+            pass
```
Comment (tancik) [inline nerfstudio/field_components/encodings.py:689]: Remove

File: nerfstudio/field_components/encodings.py [hunk 4]
```diff
@@ -647,17 +676,38 @@ class SHEncoding(Encoding):
...
+                encoding_config=encoding_config,
+            )
+
     def get_out_dim(self) -> int:
         return self.levels**2
 
     @torch.no_grad()
-    def forward(self, in_tensor: Float[Tensor, "*bs input_dim"]) -> Float[Tensor, "*bs output_dim"]:
+    def pytorch_fwd(self, in_tensor: Float[Tensor, "*bs input_dim"]) -> Float[Tensor, "*bs output_dim"]:
+        """Forward pass using pytorch. Significantly slower than TCNN implementation."""
```
Comment (tancik) [inline nerfstudio/field_components/encodings.py:707]: Is it that significant?

File: nerfstudio/field_components/mlp.py [hunk 5]
```diff
@@ -15,13 +15,39 @@
...
+        activation: torch.nn activation function
+    Returns:
+        str: TCNN activation function string
+    """
+    if isinstance(activation, nn.ReLU):
+        return "ReLU"
+    if isinstance(activation, nn.Sigmoid):
+        return "Sigmoid"
+    if isinstance(activation, type(None)):
+        return "None"
```
Comment (tancik) [inline nerfstudio/field_components/mlp.py:49]: Might as well add the others - https://github.com/NVlabs/tiny-cuda-nn/blob/master/DOCUMENTATION.md#activation-functions

File: nerfstudio/field_components/mlp.py [hunk 6]
```diff
@@ -57,7 +84,27 @@ def __init__(
         self.activation = activation
         self.out_activation = out_activation
         self.net = None
-        self.build_nn_modules()
+
+        self.tcnn_encoding = None
+        if implementation == "torch":
+            self.build_nn_modules()
+        elif implementation == "tcnn" and not TCNN_EXISTS:
+            print_tcnn_speed_warning("MLP")
```
Comment (tancik) [inline nerfstudio/field_components/mlp.py:92]: For large MLPs, this might not be true

File: nerfstudio/model_components/ray_samplers.py [hunk 7]
```diff
@@ -42,10 +42,10 @@ def __init__(
         self.num_samples = num_samples
 
     @abstractmethod
-    def generate_ray_samples(self) -> RaySamples:
+    def generate_ray_samples(self) -> Any:
```
Comment (tancik) [inline nerfstudio/model_components/ray_samplers.py:45]: Why `Any`?

File: nerfstudio/scripts/train.py [hunk 8]
```diff
@@ -158,29 +159,33 @@ def _distributed_worker(
...
     assert config is not None
-    world_size = num_machines * num_gpus_per_machine
-    if world_size <= 1:
-        # world_size=0 uses one CPU in one process.
-        # world_size=1 uses one GPU in one process.
+    world_size = num_machines * num_devices_per_machine
+    if world_size == 0:
+        raise ValueError("world_size cannot be 0")
+    elif world_size == 1:
+        # if using CPUs, then world_size=1 uses one CPU in one process
```
Comment (tancik) [inline nerfstudio/scripts/train.py:187]: -> ` # Uses one CPU or one GPU in on process`

File: nerfstudio/field_components/mlp.py (hunk 5 above)
Comment (ethanweber) [reply nerfstudio/field_components/mlp.py:49]: I added the ones for the nn.Module's and improved the error message.

File: nerfstudio/model_components/ray_samplers.py (hunk 7 above)
Comment (ethanweber) [reply nerfstudio/model_components/ray_samplers.py:45]: Due to inconsistent types in overrided functions.

**Review by tancik — APPROVED** [review]
LGTM

**Pull request merged**